common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+                       transfer_t* transfers[], size_t tx_len) {
+  size_t i, j, count, current_index = 0;
+  trit_t essence_trits[NUM_TRITS_ESSENCE];
+  trit_t bundle_trit[HASH_LENGTH_TRIT];
+  transfer_t* tf = NULL;
+  int64_t value;
+
+  // Calculate bundle hash
+  for (i = 0; i < tx_len; i++) {
+    tf = transfers[i];
+    count = transfer_transactions_count(tf);
+    for (j = 0; j < count; j++) {
+      value = tf->type == VALUE_OUT ? (j == 0 ? tf->value : 0) : 0;
+      absorb_essence(kerl, tf->address, value, tf->tag, tf->timestamp,
+                     current_index, transfer_ctx->count - 1, essence_trits);
+      current_index++;
+    }
+  }
+  // Squeeze kerl to get the bundle hash
+  kerl_squeeze(kerl, bundle_trit, HASH_LENGTH_TRIT);
+  flex_trits_from_trits(transfer_ctx->bundle, HASH_LENGTH_TRIT, bundle_trit,
+                        HASH_LENGTH_TRIT, HASH_LENGTH_TRIT);
+}
+
+transfer_iterator_t* transfer_iterator_new(transfer_t* transfers[], size_t len,
+                                           Kerl* kerl) {
+  transfer_ctx_t* transfer_ctx = NULL;
+  transfer_iterator_t* transfer_iterator =
+      (transfer_iterator_t*)calloc(1, sizeof(transfer_iterator_t));
+  if (!transfer_iterator) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  transfer_iterator->transfers = transfers;
+  transfer_iterator->transfers_count = len;
+  transfer_ctx = transfer_ctx_new();
+  if (!transfer_ctx) {
+    transfer_iterator_free(&transfer_iterator);
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  if (transfer_ctx_init(transfer_ctx, transfers, len)) {
+    transfer_iterator->transactions_count = transfer_ctx->count;
+    transfer_ctx_hash(transfer_ctx, kerl, transfers, len);
+    memcpy(transfer_iterator->bundle_hash, transfer_ctx->bundle,
+           FLEX_TRIT_SIZE_243);
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Invalid transfers.\n", __func__,
+              __LINE__);
+  }
+  transfer_ctx_free(transfer_ctx);
+  transfer_iterator->iota_signature_gen = iota_flex_sign_signature_gen;
+  return transfer_iterator;
+}
+
+void transfer_iterator_free(transfer_iterator_t** iter) {
+  transfer_iterator_t* iter_p = *iter;
+  if (*iter || iter) {
+    if (iter_p->dynamic_transaction && iter_p->transaction) {
+      free(iter_p->transaction);
+    }
+    if (iter_p->transaction_signature) {
+      free(iter_p->transaction_signature);
+    }
+    free(iter_p);
+    *iter = NULL;
+  }
+}
+
+iota_transaction_t transfer_iterator_next(
+    transfer_iterator_t* transfer_iterator) {
+  iota_transaction_t transaction = NULL;
+
+  if (transfer_iterator->current_transfer <
+      transfer_iterator->transfers_count) {
+    // Dynamically allocate a transaction structure if none was
+    // previously provided with transfer_iterator_set_transaction
+    if (!transfer_iterator->transaction) {
```

Comment:
doesn't this method suppose to be a getter/read only?
should we not return null in that case?

Reply:
what do you mean? could you give more details?

Reply:
you are allocating a new transaction, should we do that when someone calls `next`?

Reply:
it depends on the user, the comment on line #404 has answered your question.